File: solutions/go/blackjack/1/blackjack.go
package blackjack

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) int {
	switch card {
	case "ace":
		return 11
	case "ten", "jack", "queen", "king":
		return 10
	case "nine":
		return 9
	case "eight":
		return 8
	case "seven":
		return 7
	case "six":
		return 6
	case "five":
		return 5
	case "four":
		return 4
	case "three":
		return 3
	case "two":
		return 2
	default:
		return 0
	}
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {
	card1Value := ParseCard(card1)
	card2Value := ParseCard(card2)
	sum := card1Value + card2Value
	dealerCardValue := ParseCard(dealerCard)

	switch {
	case card1 == card2 && card1 == "ace":
		return "P"
	case sum == 21:
		if dealerCardValue < 10 {
			return "W"
		}
		return "S"
	case sum >= 17 && sum <= 20:
		return "S"
	case sum >= 12 && sum <= 16:
		if dealerCardValue >= 7 {
			return "H"
		}
		return "S"
	default:
		return "H"
	}
}

File: solutions/ruby/blackjack/1/blackjack.rb
module Blackjack
  def self.parse_card(card)
    case card
    when 'ace' then 11
    when 'ten', 'jack', 'queen', 'king' then 10
    when 'nine' then 9
    when 'eight' then 8
    when 'seven' then 7
    when 'six' then 6
    when 'five' then 5
    when 'four' then 4
    when 'three' then 3
    when 'two' then 2
    else 0
    end
  end

  def self.card_range(card1, card2)
    case parse_card(card1) + parse_card(card2)
    when 4..11 then 'low'
    when 12..16 then 'mid'
    when 17..20 then 'high'
    when 21 then 'blackjack'
    end
  end

  def self.first_turn(card1, card2, dealer_card)
    range = card_range(card1, card2)
    dealer_card_value = parse_card(dealer_card)

    case
    when card1 == card2 && card1 == 'ace' then 'P'
    when range == 'blackjack' then dealer_card_value < 10 ? 'W' : 'S'
    when range == 'high' then 'S'
    when range == 'mid' then dealer_card_value >= 7 ? 'H' : 'S'
    else 'H'
    end
  end
end
